kcidb/tools/kcidb_match.py

```diff
@@ -0,0 +1,471 @@
+    if snippet_lines < 0:
+        lines = text.splitlines()
+        return '\n'.join(lines[snippet_lines:])
+    return text
+
+
+class PatternDatabase:
+    """Class to handle DB table 'patterns'"""
+    def __init__(self, db_name=DB_NAME):
+        self.db_name = db_name
+        self.setup_database()
+
+    def setup_database(self):
+        """Connect to DB and create 'patterns' table if doesn't exist"""
+        with sqlite3.connect(self.db_name) as conn:
+            cursor = conn.cursor()
+            cursor.execute('''
+                CREATE TABLE IF NOT EXISTS patterns (
+                    issue_id TEXT UNIQUE,
+                    issue_version INTEGER,
+                    pattern_object JSON
+                )
+            ''')
+            conn.commit()
+
+    def add_pattern(self, issue_id, issue_version, pattern_object):
+        """Add pattern object to DB"""
+        with sqlite3.connect(self.db_name) as conn:
+            cursor = conn.cursor()
+            cursor.execute('''
+                INSERT INTO patterns (issue_id, issue_version, pattern_object)
+                VALUES (?, ?, json(?))
+                ON CONFLICT(issue_id) DO UPDATE SET
+                issue_version=excluded.issue_version,
+                pattern_object=excluded.pattern_object
+            ''', (issue_id, issue_version, json.dumps(pattern_object)))
+            conn.commit()
+
+    def remove_pattern(self, issue_id):
+        """Remove pattern object from DB"""
+        with sqlite3.connect(self.db_name) as conn:
+            cursor = conn.cursor()
+            cursor.execute('DELETE FROM patterns WHERE issue_id = ?',
+                           (issue_id,))
+            conn.commit()
+
+    def get_all_patterns(self):
+        """Retrieve all patterns objects from DB"""
+        with sqlite3.connect(self.db_name) as conn:
+            cursor = conn.cursor()
+            cursor.execute('SELECT issue_id, issue_version, pattern_object '
+                           'FROM patterns')
+            while True:
+                row = cursor.fetchone()
+                if row is None:
+                    break
+                yield row
+
+    def update_patterns(self, issue):
+        """Update patterns for existing issue"""
+        if not issue.misc.get("pattern_object"):
+            self.remove_pattern(issue.id)
+            return
+
+        pattern_object = issue.misc.get("pattern_object")
+        if not PatternValidator().validate_pattern_object(pattern_object):
+            logger.error("Pattern object validation failed for issue id: %s",
+                         issue.id)
+            return
+        self.add_pattern(issue.id, issue.version, pattern_object)
+
+
+class PatternValidator:
+    """Pattern validator class"""
+    def __init__(self):
+        self.schema = copy.deepcopy(kcidb.io.SCHEMA.json)
+        self.remove_required_fields(self.schema)
+
+    def remove_required_fields(self, schema):
```

Comment:
"""We use the same KCIDB schema to validade, but without any required fields, so remove the required rule for our validation"""